File: cat-client/src/main/java/com/dianping/cat/message/internal/DefaultTaggedTransaction.java
package com.dianping.cat.message.internal;

import com.dianping.cat.message.Message;
import com.dianping.cat.message.TaggedTransaction;
import com.dianping.cat.message.spi.MessageManager;
import com.dianping.cat.message.spi.MessageTree;

public class DefaultTaggedTransaction extends DefaultTransaction implements TaggedTransaction {
	private String m_rootMessageId;

	private String m_parentMessageId;

	private String m_tag;

	public DefaultTaggedTransaction(String type, String name, String tag, MessageManager manager) {
		super(type, name, manager);

		m_tag = tag;

		setStandalone(false);

		MessageTree tree = manager.getThreadLocalMessageTree();

		if (tree != null) {
			m_rootMessageId = tree.getRootMessageId();
			m_parentMessageId = tree.getMessageId();
		}
	}

	@Override
	public void bind(String tag, String childMessageId, String title) {
		DefaultEvent event = new DefaultEvent("RemoteCall", "Tagged");

		if (title == null) {
			title = getType() + ":" + getName();
		}

		event.addData(childMessageId, title);
		event.setTimestamp(getTimestamp());
		event.setStatus(Message.SUCCESS);
		event.setCompleted(true);

		addChild(event);
	}

	@Override
	public String getParentMessageId() {
		return m_parentMessageId;
	}

	@Override
	public String getRootMessageId() {
		return m_rootMessageId;
	}

	@Override
	public String getTag() {
		return m_tag;
	}

	@Override
	public void start() {
		MessageTree tree = getManager().getThreadLocalMessageTree();

		if (tree != null && tree.getRootMessageId() == null) {
			tree.setParentMessageId(m_parentMessageId);
			tree.setRootMessageId(m_rootMessageId);
		}
	}
}

File: lib/csharp/src/CatClient/Message/Internals/DefaultTaggedTransaction.cs
using System;
using System.Collections.Generic;
using System.Text;
using Org.Unidal.Cat.Message.Spi;

namespace Org.Unidal.Cat.Message.Internals
{
    class DefaultTaggedTransaction : DefaultTransaction, ITaggedTransaction
    {
        private string _mRootMessageId;

        private string _mParentMessageId;

        private string _mTag;

        public DefaultTaggedTransaction(string type, string name, string tag, IMessageManager manager)
            : base(type, name, manager)
        {
            _mTag = tag;
            Standalone = false;
            IMessageTree tree = manager.ThreadLocalMessageTree;
            if (tree != null)
            {
                _mRootMessageId = tree.RootMessageId;
                _mParentMessageId = tree.MessageId;
            }
        }

        public void Bind(string tag, string childMessageId, string title)
        {
            try
            {
                DefaultEvent evt = new DefaultEvent("RemoteCall", "Tagged");
                if (String.IsNullOrEmpty(title))
                {
                    title = Type + ":" + Name;
                }
                evt.AddData(childMessageId, title);
                evt.Timestamp = Timestamp;
                evt.Status = CatConstants.SUCCESS;
                evt.SetCompleted(true);
                AddChild(evt);
            }
            catch (Exception ex)
            {
                Cat.lastException = ex;
            }
        }

        public string ParentMessageId {get {return _mParentMessageId; } }

        public string RootMessageId { get { return _mRootMessageId; } }

        public string Tag { get { return _mTag; } }

        public void Start()
        {
            try
            {
                IMessageTree tree = Manager.ThreadLocalMessageTree;
                if (null != tree && tree.RootMessageId == null)
                {
                    tree.ParentMessageId = _mParentMessageId;
                    tree.RootMessageId = _mRootMessageId;
                }
            }
            catch (Exception ex)
            {
                Cat.lastException = ex;
            }
        }
    }
}
